streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredVersionedKeyValueStore.java

```diff
@@ -148,13 +205,38 @@ protected <R> QueryResult<R> runRangeQuery(final Query<R> query,
             throw new UnsupportedOperationException("Versioned stores do not support RangeQuery queries at this time.");
         }
 
+        @SuppressWarnings("unchecked")
         @Override
         protected <R> QueryResult<R> runKeyQuery(final Query<R> query,
-                                                 final PositionBound positionBound,
-                                                 final QueryConfig config) {
-            // throw exception for now to reserve the ability to implement this in the future
-            // without clashing with users' custom implementations in the meantime
-            throw new UnsupportedOperationException("Versioned stores do not support KeyQuery queries at this time.");
+            final PositionBound positionBound,
+            final QueryConfig config) {
+            if (query instanceof VersionedKeyQuery) {
+                final QueryResult<R> result;
+                final VersionedKeyQuery<K, V> typedKeyQuery = (VersionedKeyQuery<K, V>) query;
+                VersionedKeyQuery<Bytes, byte[]> rawKeyQuery;
+                if (typedKeyQuery.asOfTimestamp().isPresent()) {
+                    rawKeyQuery = VersionedKeyQuery.withKey(keyBytes(typedKeyQuery.key()));
+                    rawKeyQuery = rawKeyQuery.asOf(typedKeyQuery.asOfTimestamp().get());
+                } else {
+                    rawKeyQuery = VersionedKeyQuery.withKey(keyBytes(typedKeyQuery.key()));
+                }
+                final QueryResult<VersionedRecord<byte[]>> rawResult =
+                    wrapped().query(rawKeyQuery, positionBound, config);
+                if (rawResult.isSuccess()) {
+                    final Function<byte[], ValueAndTimestamp<V>> deserializer = getDeserializeValue(plainValueSerdes);
```

Comment:
Why is the type `ValueAndTimestamp`? Should it not be `VersionedRecord`?

Reply:
> Why is the type `ValueAndTimestamp`? Should it not be `VersionedRecord`?

Thanks Matthias. Good point. I wanted to discuss it with you.
I just did it to reuse the `ValueAndTimestampDeserializer` class. Moreover, inside `MeteredVersionedKeyValueStoreInternal` every other method uses `ValueAndTimestamp` as well.
Btw, do you agree to implement a new class `VersionedRecordDeserializer`?

Reply:
> MeteredVersionedKeyValueStoreInternal every other method uses ValueAndTimestamp as well.

This part is for DSL compatibility only. All DSL operator only work with `ValueAndTimestamp` type right now, so when we plugin the new versioned state store, we need to translate between the types to make it work.

> Btw, do you agree to implement a new class VersionedRecordDeserializer?

As this point, I don't think we need such a class? The helper method you added does the job perfectly fine IMHO. But I am also not against it, if it make it easier (we would add it as internal class though only I assume?).